### PR TITLE
Restore Docker autobuilds on repo pushes via GitHub Actions

### LINKED ISSUE #18: Docker Autobuilds are disabled
@guolinke I got the following e-mail:

---

Action Required: Upgrade to continue using Docker Autobuilds
===

Hi there!

As you may remember, on June 7, 2021 we [announced](https://www.docker.com/blog/changes-to-docker-hub-autobuilds) the end of Autobuilds for all free accounts and starting on Monday, July 26, 2021 all Autobuilds associated with a free account will be disabled and stop working.

If you would like to ensure your currently setup Autobuilds continue to function beyond the July 26th shutoff date please upgrade to a [Pro or Team plan](https://www.docker.com/pricing).

Thanks,
The Docker Team

---

I think we can continue using Docker Hub for this repo but have to trigger all builds (for PRs and `master` branch) manually. Given the frequency of commits in this repo, I don't think this will be hard.

### PR DESCRIPTION
**Pull request by StrikerRUS**
Closed #18.

GitHub Actions logs: https://github.com/guolinke/lightgbm-ci-docker/runs/3641088439?check_suite_focus=true.

Push to Docker Hub succeeded: https://hub.docker.com/layers/40251340/lightgbm/vsts-agent/ubuntu-14.04-dev/images/sha256-10e440ae8683e5a0694bd4978bc0ce77a2f5ab0ff176a1540e75065eeeec6edc?context=repo.

image

Azure builds were successful with recently uploaded Docker file:

image

image

cc @jameslamb

### CODE REVIEW

File: .github/workflows/publish_image.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    - master
+    - dev
+
+jobs:
+  push_to_docker_hub:
+    name: Push Docker image to Docker Hub
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2.3.4
```
Comment (jameslamb) [inline .github/workflows/publish_image.yml]:
```suggestion
        uses: actions/checkout@v2
```

The `checkout` action maintains a tag for each major release, which is updated on the creation of each new minor or patch release.

https://github.com/actions/checkout/tags

What do you think about just pinning to `v2`? That way, no new work would be required in this repo if they publish a v2.3.5 with some bugfixes.

**Review by jameslamb — APPROVED** [review]
thanks for this!

File: .github/workflows/publish_image.yml (hunk 1 above)
Comment (StrikerRUS) [reply .github/workflows/publish_image.yml]: Ah, I didn't notice that! Thanks a lot for the suggestion! It turned out that the same is true for `docker/***` actions, btw.

File: .github/workflows/publish_image.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Log in to Docker Hub
+        uses: docker/login-action@v1.10.0
```
Comment (StrikerRUS) [inline .github/workflows/publish_image.yml]:
```suggestion
        uses: docker/login-action@v1
```

File: .github/workflows/publish_image.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Extract tag for Docker
+        id: tag
+        shell: bash
+        run: |
+            branch=${GITHUB_REF#refs/heads/}
+            docker_tag=ubuntu-14.04-${branch}
+            docker_tag=${docker_tag%-master}
+            echo "::set-output name=docker_tag::${docker_tag}"
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v2.7.0
```
Comment (StrikerRUS) [inline .github/workflows/publish_image.yml]:
```suggestion
        uses: docker/build-push-action@v2
```

**Comment by StrikerRUS** [comment]
@guolinke Is this PR OK for you?

**Pull request merged**